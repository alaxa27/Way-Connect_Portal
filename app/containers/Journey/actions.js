/*
 *
 * Journey actions
 *
 */

import {
  DEFAULT_ACTION,
  JOURNEY_ID_OUTOFRANGE,
  JOURNEY_ID_INCREASED,
  JOURNEY_ID_DECREASED,
  CHANGE_DEFAULT_ANSWERS_LIST,
  ACKNOWLEDGE_COMMUNICATION,
  ACKNOWLEDGE_COMMUNICATION_SUCCESS,
  ACKNOWLEDGE_COMMUNICATION_ERROR,
  SKIP_VIDEO,
  SKIP_VIDEO_SUCCESS,
  SKIP_VIDEO_ERROR,
  COMPLETE_JOURNEY,
  COMPLETE_JOURNEY_SUCCESS,
  COMPLETE_JOURNEY_ERROR,
  AUTHENTICATE,
  AUTHENTICATE_SUCCESS,
  AUTHENTICATE_ERROR,
  ANSWER_QUESTION,
  ANSWER_QUESTION_SUCCESS,
  ANSWER_QUESTION_ERROR,
} from './constants';

export function defaultAction() {
  return {
    type: DEFAULT_ACTION,
  };
}

export function changeID(prevID, currentID, length) {
  if (currentID >= length) return { type: JOURNEY_ID_OUTOFRANGE, currentID };
  if (prevID < currentID) {
    return {
      type: JOURNEY_ID_INCREASED,
      currentID,
    };
  }
  return {
    type: JOURNEY_ID_DECREASED,
    currentID,
  };
}

export function changeDefaultAnswersList(defaultAnswers, questionID) {
  return {
    type: CHANGE_DEFAULT_ANSWERS_LIST,
    defaultAnswers,
    questionID,
  };
}

export function acknowledgeCommunication() {
  return {
    type: ACKNOWLEDGE_COMMUNICATION,
  };
}

export function communicationAcknowledged() {
  return {
    type: ACKNOWLEDGE_COMMUNICATION_SUCCESS,
  };
}

export function acknowledgingCommunicationError() {
  return {
    type: ACKNOWLEDGE_COMMUNICATION_ERROR,
  };
}

export function skipVideo() {
  return {
    type: SKIP_VIDEO,
  };
}

export function videoSkipped() {
  return {
    type: SKIP_VIDEO_SUCCESS,
  };
}

export function skippingVideoError() {
  return {
    type: SKIP_VIDEO_ERROR,
  };
}

export function completeJourney() {
  return {
    type: COMPLETE_JOURNEY,
  };
}

export function journeyCompleted(data) {
  window.location.href = data.url;
  return {
    type: COMPLETE_JOURNEY_SUCCESS,
  };
}

export function journeyCompletionError() {
  return {
    type: COMPLETE_JOURNEY_ERROR,
  };
}

export function authenticate() {
  return {
    type: AUTHENTICATE,
  };
}

export function authenticated(data) {
  window.location.href = data.url;
  return {
    type: AUTHENTICATE_SUCCESS,
  };
}

export function authenticationError() {
  return {
    type: AUTHENTICATE_ERROR,
  };
}

export function answerQuestion() {
  return {
    type: ANSWER_QUESTION,
  };
}

export function questionAnswered() {
  return {
    type: ANSWER_QUESTION_SUCCESS,
  };
}

export function answeringQuestionError() {
  return {
    type: ANSWER_QUESTION_ERROR,
  };
}
