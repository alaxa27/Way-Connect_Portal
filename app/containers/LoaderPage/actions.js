/*
 *
 * LoaderPage actions
 *
 */

import {
  DEFAULT_ACTION,
  GET_ESTABLISHMENT,
  GET_ESTABLISHMENT_SUCCESS,
  GET_ESTABLISHMENT_ERROR,
  POST_CONNECTION,
  POST_CONNECTION_SUCCESS,
  POST_CONNECTION_ERROR,
  GET_PROMOTION_LEVELS,
  GET_PROMOTION_LEVELS_SUCCESS,
  GET_PROMOTION_LEVELS_ERROR,
  RETRIEVE_DISCOUNT,
  RETRIEVE_DISCOUNT_SUCCESS,
  RETRIEVE_DISCOUNT_ERROR,
} from './constants';

export function defaultAction() {
  return {
    type: DEFAULT_ACTION,
  };
}

export function loadEstablishment() {
  return {
    type: GET_ESTABLISHMENT,
  };
}

export function establishmentLoaded(name, picture) {
  return {
    type: GET_ESTABLISHMENT_SUCCESS,
    name,
    picture,
  };
}

export function establishmentLoadingError(error) {
  return {
    type: GET_ESTABLISHMENT_ERROR,
    error,
  };
}

export function postConnection() {
  return {
    type: POST_CONNECTION,
  };
}

export function connectionPosted(connection) {
  return {
    type: POST_CONNECTION_SUCCESS,
    communication: connection.communication,
  };
}

export function connectionPostingError() {
  return {
    type: POST_CONNECTION_ERROR,
  };
}

export function loadPromotionLevels() {
  return {
    type: GET_PROMOTION_LEVELS,
  };
}

export function promotionLevelsLoaded(promotionLevels) {
  const removeAndGetText = (rank, levelsArray) => {
    const index = levelsArray.findIndex(level => level.rank === rank);
    if (index > -1) {
      const res = levelsArray[index].text;
      levelsArray.splice(index, 1);
      return res;
    }
    return '';
  };
  const bannerText = removeAndGetText(101, promotionLevels);
  const claimPhoneNumber = removeAndGetText(102, promotionLevels);
  return {
    type: GET_PROMOTION_LEVELS_SUCCESS,
    promotionLevels,
    bannerText,
    claimPhoneNumber,
  };
}

export function promotionLevelsLoadingError() {
  return {
    type: GET_PROMOTION_LEVELS_ERROR,
  };
}

export function loadDiscount() {
  return {
    type: RETRIEVE_DISCOUNT,
  };
}

export function discountLoaded(discount) {
  return {
    type: RETRIEVE_DISCOUNT_SUCCESS,
    discount,
  };
}

export function discountLoadingError() {
  return {
    type: RETRIEVE_DISCOUNT_ERROR,
  };
}
