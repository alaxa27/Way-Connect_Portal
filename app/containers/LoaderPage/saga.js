import { call, put, takeLatest, take, all } from 'redux-saga/effects';
import { push } from 'connected-react-router';
import axios from 'axios';

import {
  GET_ESTABLISHMENT,
  POST_CONNECTION_SUCCESS,
  RETRIEVE_DISCOUNT_SUCCESS,
  GET_PROMOTION_LEVELS_SUCCESS,
} from './constants';
import {
  establishmentLoaded,
  establishmentLoadingError,
  connectionPosted,
  connectionPostingError,
  promotionLevelsLoaded,
  promotionLevelsLoadingError,
  discountLoaded,
  discountLoadingError,
} from './actions';

const macAddress = 'AA:AD:4D:F2:1B:1B';
const apiURL = 'http://localhost:5000';

function getEstablishmentRequest() {
  return axios({
    method: 'get',
    url: `${apiURL}/customers/establishment/`,
  });
}

function postConnectionRequest() {
  return axios({
    method: 'post',
    url: `${apiURL}/customers/${macAddress}/connect/`,
  });
}

function getDiscountRequest() {
  return axios({
    method: 'get',
    url: `${apiURL}/customers/${macAddress}/retrieve_discount/`,
  });
}

function getPromotionLevelsRequest() {
  return axios({
    method: 'get',
    url: `${apiURL}/customers/levels/`,
  });
}

export function* getEstablishmentEffect() {
  try {
    const { data } = yield call(getEstablishmentRequest);
    yield put(establishmentLoaded(data.name));
  } catch (err) {
    yield put(establishmentLoadingError(err));
  }
}

export function* postConnectionEffect() {
  try {
    const { data } = yield call(postConnectionRequest);
    yield put(connectionPosted(data));
  } catch (err) {
    yield put(connectionPostingError(err));
  }
}

export function* getPromotionLevelsEffect() {
  try {
    const { data } = yield call(getPromotionLevelsRequest);
    yield put(promotionLevelsLoaded(data));
  } catch (err) {
    yield put(promotionLevelsLoadingError(err));
  }
}

export function* getDiscountEffect() {
  try {
    const { data } = yield call(getDiscountRequest);
    yield put(discountLoaded(data));
  } catch (err) {
    yield put(discountLoadingError(err));
  }
}

// Individual exports for testing
export default function* loaderPageSaga() {
  yield all([
    takeLatest(GET_ESTABLISHMENT, getEstablishmentEffect),
    takeLatest(GET_ESTABLISHMENT, postConnectionEffect),
    takeLatest(POST_CONNECTION_SUCCESS, getDiscountEffect),
    takeLatest(POST_CONNECTION_SUCCESS, getPromotionLevelsEffect),
  ]);
  yield all([
    take(RETRIEVE_DISCOUNT_SUCCESS),
    take(GET_PROMOTION_LEVELS_SUCCESS),
  ]);
  yield put(push('/journey/0'));
}
