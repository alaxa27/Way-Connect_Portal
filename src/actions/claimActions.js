// Actions are payloads of information that send data from your application to your store.
// They are the only source of information for the store.

// Actions are triggered either by user through interactions or by an events, such as successful AJAX call.

// Read more on Actions - https://redux.js.org/docs/basics/Actions.html
import axios from "axios";

import { axiosInstance } from "../constants/ApiConfig.js";
import {
  POST_CLAIM,
  POST_CLAIM_FULFILLED,
  POST_CLAIM_REJECTED
} from "../constants/ActionTypes";

export function postClaim(payload) {
  return async (dispatch, getState) => {
    if (!getState().claim.posting) {
      dispatch({
        type: POST_CLAIM
      });
      try {
        const informationData = {
          ...getState().information.informationData
        };
        const claimData = {
          ...getState().claim.claimData
        };
        const response = await axiosInstance({
          method: "post",
          url: `/customers/claim/`,
          data: {
            ...claimData,
            mac_address: informationData.mac_address
          }
        });

        dispatch({
          type: POST_CLAIM_FULFILLED,
          payload: response.data.video
        });

      } catch (error) {
        dispatch({
          type: POST_CLAIM_REJECTED
        });
        console.error(error);
      }
    }
  };
}
