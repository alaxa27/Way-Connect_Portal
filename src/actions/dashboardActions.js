// Actions are payloads of information that send data from your application to your store.
// They are the only source of information for the store.

// Actions are triggered either by user through interactions or by an events, such as successful AJAX call.

// Read more on Actions - https://redux.js.org/docs/basics/Actions.html
import {
  axiosInstance
} from "../constants/ApiConfig.js";

import {
  ACKNOWLEDGE_COMMUNICATION,
  ACKNOWLEDGE_COMMUNICATION_FULFILLED,
  ACKNOWLEDGE_COMMUNICATION_REJECTED
} from "../constants/ActionTypes";

export function acknowledgeCommunication(payload) {
  return async (dispatch, getState) => {
    dispatch({
      type: ACKNOWLEDGE_COMMUNICATION
    });
    try {
      const informationData = { ...getState().information.informationData
      };
      const response = await axiosInstance({
        method: "post",
        url: "/customers/acknowledge_communication/",
        data: {
          mac_address: informationData.mac_address
        }
      });

      dispatch({
        type: ACKNOWLEDGE_COMMUNICATION_FULFILLED,
      });
      payload.history.push("/video");

      // window.location.href = `http://192.168.220.2:2050/nodogsplash_auth/?tok=${informationData.token}&redir=${informationData.redir}`;

    } catch (error) {
      dispatch({
        type: ACKNOWLEDGE_COMMUNICATION_REJECTED
      });
    }
  };
}
