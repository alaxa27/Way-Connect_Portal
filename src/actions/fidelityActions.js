// Actions are payloads of information that send data from your application to your store.
// They are the only source of information for the store.

// Actions are triggered either by user through interactions or by an events, such as successful AJAX call.

// Read more on Actions - https://redux.js.org/docs/basics/Actions.html

import {
  FETCH_FIDELITY,
  FETCH_FIDELITY_FULFILLED,
  FETCH_FIDELITY_REJECTED,
  FETCH_DISCOUNT,
  FETCH_DISCOUNT_FULFILLED,
  FETCH_DISCOUNT_REJECTED
} from "../constants/ActionTypes";

export function fetchFidelity(payload) {
  return (dispatch, getState) => {
    dispatch({
      type: FETCH_FIDELITY,
    })
    setTimeout(function() {
      let data = getState().fidelity.fidelityData
      data.rate = 0.8
      dispatch({
        type: FETCH_FIDELITY_FULFILLED,
        payload: data
      });
    }, 2000);
  }
}

export function fetchDiscount(payload) {
  return (dispatch, getState) => {
    if (!getState().fidelity.fetching && !getState().fidelity.fetched) {
      dispatch({
        type: FETCH_DISCOUNT
      })
      setTimeout(function() {
        let data = getState().fidelity.fidelityData
        data.rate = 0;
        dispatch({
          type: FETCH_DISCOUNT_FULFILLED,
          payload: data
        });
      }, 1000);
    }
  }
}
