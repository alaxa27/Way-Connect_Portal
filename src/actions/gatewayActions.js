// Actions are payloads of information that send data from your application to your store.
// They are the only source of information for the store.

// Actions are triggered either by user through interactions or by an events, such as successful AJAX call.

// Read more on Actions - https://redux.js.org/docs/basics/Actions.html
import {
  axiosInstance
} from "../constants/ApiConfig.js";

import {
  ACKNOWLEDGE_COMMUNICATION,
  ACKNOWLEDGE_COMMUNICATION_FULFILLED,
  ACKNOWLEDGE_COMMUNICATION_REJECTED,

  CLICK_COMMUNICATION,
  CLICK_COMMUNICATION_FULFILLED,
  CLICK_COMMUNICATION_REJECTED,

  FETCH_ESTABLISHMENT,
  FETCH_ESTABLISHMENT_FULFILLED,
  FETCH_ESTABLISHMENT_REJECTED,
} from "../constants/ActionTypes";

export function fetchEstablishment(payload) {
  return async (dispatch, getState) => {
    dispatch({
      type: FETCH_ESTABLISHMENT
    });
    try {
      const informationData = { ...getState().information.informationData
      };
      const response = await axiosInstance({
        method: "get",
        url: "/customers/establishment/",
      });

      dispatch({
        type: FETCH_ESTABLISHMENT_FULFILLED,
        payload: response.data
      });
    } catch (error) {
      dispatch({
        type: FETCH_ESTABLISHMENT_REJECTED
      });
    }
  };
}

export function acknowledgeCommunication(payload) {
  return async (dispatch, getState) => {
    dispatch({
      type: ACKNOWLEDGE_COMMUNICATION
    });
    try {
      const informationData = { ...getState().information.informationData
      };
      const response = await axiosInstance({
        method: "post",
        url: `/customers/${informationData.mac_address}/acknowledge_communication/`,
        data: {}
      });

      dispatch({
        type: ACKNOWLEDGE_COMMUNICATION_FULFILLED,
      });
      // payload.history.push("/dashboard");

      // window.location.href = `http://192.168.220.2:2050/nodogsplash_auth/?tok=${informationData.token}&redir=${informationData.redir}`;

    } catch (error) {
      dispatch({
        type: ACKNOWLEDGE_COMMUNICATION_REJECTED
      });
    }
  };
}

export function clickCommunication(payload) {
  return async (dispatch, getState) => {
    dispatch({
      type: CLICK_COMMUNICATION
    });
    try {
      const informationData = { ...getState().information.informationData
      };
      const response = await axiosInstance({
        method: "post",
        url: `/customers/${informationData.mac_address}/click_communication/`,
        data: {}
      });

      var tempLink = document.createElement("a");
      tempLink.style.display = "none";
      tempLink.href = informationData.redirection;
      tempLink.click();
      payload.history.push("/dashboard");
      document.body.removeChild(tempLink);

      dispatch({
        type: CLICK_COMMUNICATION_FULFILLED,
      });


      // window.location.href = `http://192.168.220.2:2050/nodogsplash_auth/?tok=${informationData.token}&redir=${informationData.redir}`;

    } catch (error) {
      dispatch({
        type: CLICK_COMMUNICATION_REJECTED
      });
    }
  };
}
