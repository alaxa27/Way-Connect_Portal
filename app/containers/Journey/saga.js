import { takeEvery, call, put, select, all } from 'redux-saga/effects';
import { getDiscountEffect } from 'containers/LoaderPage/saga';
import axiosInstance from '../../apiConfig';
import {
  journeyCompleted,
  journeyCompletionError,
  communicationAcknowledged,
  acknowledgingCommunicationError,
  authenticated,
  authenticationError,
  questionAnswered,
  answeringQuestionError,
} from './actions';
import {
  makeSelectJourneyItem,
  makeSelectPreviousID,
  makeSelectCurrentID,
} from './selectors';
import { JOURNEY_ID_CHANGED } from './constants';

function completeJourneyRequest() {
  return axiosInstance({
    method: 'post',
    url: '/customers/complete/',
  });
}

function authenticateRequest() {
  return axiosInstance({
    method: 'get',
    url: '/customers/authenticate',
  });
}

function acknowledgeCommunicationRequest() {
  return axiosInstance({
    method: 'post',
    url: `/campaigns/communications/videos/acknowledge/`,
  });
}

function answerQuestionRequest(data) {
  return axiosInstance({
    method: 'post',
    url: '/questions/answers/',
    data,
  });
}

// function skipVideoRequest() {
//   return axiosInstance({
//     method: 'post',
//     url: `/campaigns/communications/videos/acknowledge/`,
//   });
// }

// export function* redirectEffect() {
//   window.location.href = 'http://google.com/';
// }

export function* completeJourneyEffect() {
  try {
    const { data } = yield call(completeJourneyRequest);
    yield put(journeyCompleted(data));
  } catch (err) {
    yield put(journeyCompletionError(err));
  }
}

export function* authenticateEffect() {
  try {
    const { data } = yield call(authenticateRequest);
    yield put(authenticated(data));
  } catch (err) {
    yield put(authenticationError(err));
  }
}

export function* answerQuestionEffect(id, type, answers) {
  const answer = {
    question: id,
  };

  switch (type) {
    case 'VALUE':
      // eslint-disable-next-line no-case-declarations
      const value1 = answers[0];
      answer.value = value1;
      break;
    case 'CHOICE':
      answer.choices = answers;
      break;
    default:
      break;
  }

  try {
    yield call(answerQuestionRequest, answer);
    yield put(questionAnswered());
  } catch (err) {
    yield put(answeringQuestionError(err));
  }
}

// export function* skipVideoEffect() {
//   try {
//     yield call(skipVideoRequest);
//     yield put(videoSkipped());
//   } catch (err) {
//     yield put(skippingVideoError(err));
//   }
// }

export function* acknowledgeCommunicationEffect() {
  try {
    yield call(acknowledgeCommunicationRequest);
    yield put(communicationAcknowledged());
  } catch (err) {
    yield put(acknowledgingCommunicationError(err));
  }
}

export function* handleCurrentEffect(journeyItem) {
  console.log('current', journeyItem.toJS());
  switch (journeyItem.type) {
    case 'C':
      yield call(acknowledgeCommunicationEffect);
      yield call(getDiscountEffect);
      break;
    case 'END':
      yield all([call(authenticateEffect), call(completeJourneyEffect)]);
      break;
    default:
      break;
  }
  return null;
}

export function* handlePreviousEffect(journeyItem) {
  console.log('previous', journeyItem.toJS());
  switch (journeyItem.type) {
    case 'Q':
      yield call(
        answerQuestionEffect,
        journeyItem.question.id,
        journeyItem.question.type,
        journeyItem.question.defaultAnswers,
      );
      break;
    default:
      break;
  }
  return null;
}

export function* journeyIDChangedEffect() {
  const previousIDSelector = makeSelectPreviousID();
  const currentIDSelector = makeSelectCurrentID();

  const previousID = yield select(previousIDSelector);
  const currentID = yield select(currentIDSelector);
  console.log(previousID);
  console.log(currentID);

  const previousJourneyItemSelector = makeSelectJourneyItem(previousID);
  const currentJourneyItemSelector = makeSelectJourneyItem(currentID);

  const currentJourneyItem = yield select(currentJourneyItemSelector);
  const previousJourneyItem = yield select(previousJourneyItemSelector);

  yield all([
    call(handlePreviousEffect, previousJourneyItem),
    call(handleCurrentEffect, currentJourneyItem),
  ]);
}

// Individual exports for testing
export default function* journeySaga() {
  yield takeEvery(JOURNEY_ID_CHANGED, journeyIDChangedEffect);
}
