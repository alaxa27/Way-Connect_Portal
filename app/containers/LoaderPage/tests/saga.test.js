/**
 * Test sagas
 */

/* eslint-disable redux-saga/yield-effects */
// import { take, call, put, select } from 'redux-saga/effects';
import { put } from 'redux-saga/effects';

// import { GET_ESTABLISHMENT } from '../constants';

import { establishmentLoaded, establishmentLoadingError } from '../actions';
import { getEstablishmentEffect } from '../saga';

// const generator = loaderPageSaga();

describe('getEstablishment effect', () => {
  let getEstablishmentGenerator;

  beforeEach(() => {
    getEstablishmentGenerator = getEstablishmentEffect();

    const callDescriptor = getEstablishmentGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the establishmentLoaded action if the request is successful', () => {
    const data = {
      name: 'foo',
    };
    const putDescriptor = getEstablishmentGenerator.next({ data }).value;
    // console.log(putDescriptor)
    expect(putDescriptor).toEqual(put(establishmentLoaded(data.name)));
  });

  it('should dispatch the establishmentLoadingError action requests fails', () => {
    const response = new Error('Some error');
    const putDescriptor = getEstablishmentGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(establishmentLoadingError(response)));
  });
});
// describe('loaderPageSaga Saga', () => {
//   const loaderPageSaga = loaderPage();

//   it('should start task to watch for GET_ESTABLISHMENT action', () => {
//     const takeLatestDescriptor = loaderPageSaga.next().value;
//     expect(takeLatestDescriptor).toEqual(
//       takeLatest(GET_ESTABLISHMENT, getEstablishmentEffect),
//     );
//   });
// });
