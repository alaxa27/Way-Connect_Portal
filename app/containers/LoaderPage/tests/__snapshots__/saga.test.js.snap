// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getEstablishment effect should dispatch the establishmentLoaded action if the request is successful 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`getEstablishment effect should dispatch the establishmentLoadingError action requests fails 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
}
`;
